Only create a target for object descriptions

latosha-maltba opened the pull request:
Subject: Only create a target for object descriptions

### Feature or Bugfix
- Feature

### Purpose
Adding feature for https://github.com/sphinx-doc/sphinx/issues/9662 (ability to only create a target for a domain description).
Thanks again to @jakobandersen for the inspiration.

This is #9675 repurposed and rebased onto the current 5.x. In the hope that it now gets more traction.

Open Tasks:
* Bikeshedding:  should the parameter be named `:hidden:`,  `:targetonly:`, `:your-favorite-word-here:`, ...
* If this implementation gets an approval, I'll add the documentation for the parameter before merging.

### Relates
* Closes https://github.com/sphinx-doc/sphinx/issues/9662
* https://github.com/sphinx-doc/sphinx/pull/9671
* https://github.com/sphinx-doc/sphinx/pull/9675

AA-Turner commented on sphinx/directives/__init__.py:
```diff
@@ -203,8 +205,30 @@ def run(self) -> List[Node]:
         return [self.indexnode, node]
 
+    # TODO: This method does not need access to self, make a (non-class) function out of it?
+    def replace_node_with_target(self, node: nodes.Node) -> nodes.target:
```
```suggestion
    @staticmethod
    def replace_node_with_target(node: nodes.Node) -> nodes.target:
```

AA-Turner commented on sphinx/directives/__init__.py:
```diff
@@ -203,8 +205,30 @@ def run(self) -> List[Node]:
+                    ids.extend(collect_ids(c))
+                return ids
+            else:
+                return []
```
Does this function need to be defined as a closure? If not please factor out for speed.

AA-Turner commented on sphinx/directives/__init__.py:
```diff
@@ -203,8 +205,30 @@ def run(self) -> List[Node]:
+        ids: List[str] = collect_ids(node)
+        target_node = nodes.target()
+        target_node['ids'] = ids
+        return target_node
```
Does this work?
```suggestion
        target_node = nodes.target(ids=collect_ids(node))
        return target_node
```

AA-Turner commented on sphinx/domains/javascript.py:
```diff
@@ -37,10 +37,10 @@ class JSObject(ObjectDescription[Tuple[str, str]]):
-    option_spec: OptionSpec = {
-        'noindex': directives.flag,
+    option_spec: OptionSpec = ObjectDescription.option_spec.copy()
+    option_spec.update({
```
Please just copy `:hiden:` to all subclasses, it makes it easier to reason about.

AA-Turner replied:
Makes sense, the current method (`.update`) is fine.

A

AA-Turner reviewed (changes requested):
This needs a lot of tests -- for the new transform and for the behaviour x each domain.

A

jakobandersen commented on sphinx/domains/cpp.py:
```diff
@@ -7083,10 +7083,12 @@ class CPPObject(ObjectDescription[ASTDeclaration]):
     ]
 
-    option_spec: OptionSpec = {
+    option_spec: OptionSpec = ObjectDescription.option_spec.copy()
```
Can you change the ``CPPAlias`` class similarly to the ``CAlias`` class? I.e., copy this ``option_spec`` and delete  ``notypesetting``.

jakobandersen commented on doc/usage/restructuredtext/domains.rst:
```diff
@@ -50,13 +50,18 @@ give the directive option flag ``:noindexentry:``.
 for cross-referencing, you can give the directive option flag ``:noindex:``
 (which implies ``:noindexentry:``).
+If you do not want to typeset anything but create a target and index entry for
+later reference, you can give the directive option flag ``:notypesetting:``.
```
This may be slightly ambiguous: does ``notypesetting`` imply not ``noindex``/``noindex``? I would assume that it is independent. I.e., if the user specifies ``notypesetting`` and ``noindexentry``/``noindex``, then there will be almost no side-effect of having the directive.
```suggestion
If you do not want to typeset anything you can give the directive option flag ``:notypesetting:``.
This can for example be used to create only a target and index entry for later reference.
```

jakobandersen reviewed (changes requested):
Looks good to me. I like ``notypesetting`` as it conveys exactly what it is doing, and it is in line with ``noindex`` and ``noindexentry``.

Commit:
ObjectDescription: Add option :hidden:

Add option :hidden: to ObjectDescription.  Currently, this option has no
effect except being stored as attribute in the resulting node.

Commit:
Domain Python: Fix tests due to new index/target node order

The post transform ReorderConsecutiveTargetAndIndexNodes reorders index
and target nodes.  A snipped like::

   .. py:module:: mymodule
   .. py:data:: myvar

generates the nodes::

   <index mymodule>
   <target mymodule>
   <index myvar>
   <descr myvar>

which get reordered to::

   <index mymodule>
   <index myvar>
   <target mymodule>
   <descr myvar>

Commit:
Domain CPP: Update `option_spec` of superclass instead replacing it

Instead of overriding the `option_spec` variable copy the `option_spec`
of the superclass and update it.  This way changes in the superclass
propagate to its children.  This is in line how other classes use
"`option_spec`-inheritance".

Commit:
Domain JavaScript: Fix order of index and target nodes

Correct node order generated by the .. js:module:: directive:  When
generating an index entry, the index node must come before the target
node.

Commit:
Document :notypesetting: directive option for domains

Commit:
ObjectDescription: Replace collect_ids() with pure function

The collect_ids() function uses the its second parameter ``ids:
List[str]`` as in- and output parameter and always returns None.
Rewrite the function to become pure and side effect free: return the
list of ids as the return value not as an output parameter.

Commit:
Reorder consecutive index and target nodes

An index directive creates a index node followed by a target node.  Two
consecutive index directives::

.. index:: first
.. index:: second

create index, target, index, target, i.e. a mixture.  The interspersed
index nodes prevent other transformations, e.g. PropagateTargets to
properly work on the target nodes.

Apply a transformation which reorders mixed and consecutive index and
target nodes such that first all index nodes are before all target
nodes:

Given the following document as input:

    <document>
        <target ids="id1" ...>
        <index entries="...1...">
        <target ids="id2" ...>
        <target ids="id3" ...>
        <index entries="...2...">
        <target ids="id4" ...>

The transformed result will be:

    <document>
        <index entries="...1...">
        <index entries="...2...">
        <target ids="id1" ...>
        <target ids="id2" ...>
        <target ids="id3" ...>
        <target ids="id4" ...>

Commit:
Domain Python: Fix order of index and target nodes

Correct node order generated by the .. py:module:: directive:  When
generating an index entry, the index node must come before the target
node.

Commit:
ObjectDescription: Hide contents if :hidden: is given

Do not produce any output only a target node if option :hidden: is
given.  In this case all nodes representing this ObjectDescription are
replaced by a single target where the target gets assigned all ids of
the replaced nodes.

Commit:
ObjectDescription: Record source location for targets

Record the source information (class attributes source and line) when
replacing an object description node with a target node.

Commit:
ObjectDescription: Factor out collect_ids() method

Move the collect_ids() method from inside the replace_node_with_target()
into the class scope (as static method) since it does not need the
closure of the surrounding replace_node_with_target() method.  This also
gives a potential speed boost.

Commit:
ObjectDescription: Rename :hidden: to :notypesetting:

Commit:
ObjectDescription: Simplify replace_node_with_target()

Commit:
Domain C and CPP: Drop ``:notypesetting:`` from alias object

The alias objects override the run() method and thus do not support
``:notypesetting:`` option.

Commit:
ObjectDescription: Make replace_node_with_target() a static method

replace_node_with_target() does not access the self-instance.  Thus,
declare it as a static method.

Commit:
Add tests for ReorderConsecutiveTargetAndIndexNodes

Commit:
ObjectDescription: Add tests for :notypesetting: option

Commit:
ObjectDescription: Inline replace_node_with_target()

Inline the one-line method replace_node_with_target() into its only call
site.  Since there are no call sites left, remove the method afterwards.

Commit:
CHANGES: Add entry for ``:notypesetting:``

Commit:
ObjectDescription: Do not create a target node without ids

It might be that a object description has no ids associated with it,
e.g. if the option :noindex: is passed.  In this case we would replace
the object description node with a target node, which has no ids.  This
breaks docutils assumption about a target node and leads not errors.
Therefore, do only create a target node if we have ids to work with.

AA-Turner committed:
Merge branch 'master' into object_description_run_5.x

# Conflicts:
#	doc/usage/restructuredtext/domains.rst
#	sphinx/directives/__init__.py
#	sphinx/domains/c.py
#	sphinx/domains/cpp.py
#	sphinx/domains/javascript.py
#	sphinx/domains/python.py
#	sphinx/transforms/__init__.py

AA-Turner committed:
typing

AA-Turner committed:
style

AA-Turner committed:
notypesetting -> no-typesetting

AA-Turner committed:
Don't copy ``ObjectDescription.option_spec``

AA-Turner committed:
Add to reST domain

AA-Turner committed:
Add to JSModule and PythonModule

AA-Turner committed:
Updates

AA-Turner committed:
Quotation marks

AA-Turner committed:
rename

AA-Turner committed:
rename

AA-Turner committed:
Update CHANGES entry.
